### PR TITLE
Plugin/RELOAD - Tune usage of var global, add limit to options

### PR DESCRIPTION
**Pull request by fturib**
### 1. What does this pull request do?
Tune the plugin RELOAD
- found that the RELOAD was going inactive after load of invalid corefile. That may depend where is the error on the corefile (I mean if the setup for RELOAD was fully completed or not)
- add limits to the parameter to avoid issues with 0.

### 2. Which issues (if any) are related?
fix partially #1455 

### 3. Which documentation changes (if any) need to be made?
modified the README for minimal values. 

**IMPORTANT**: There is still an issue at shutdown: with no RELOAD, the server stop nicely, with several reload (and thus restart), the server does not stop nicely. I guess it needs to revisit the go function that runs in background and that should stop. Maybe link to #1456

### CODE REVIEW

**Comment by fturib** [comment]
Why is the Codecov failing on a commit that does not contain code ?

**Comment by miekg** [comment]
Why don't we just send the reload signal to ourself?

**Comment by miekg** [comment]
This seems to work fine (note: little testing)
~~~
diff --git a/plugin/reload/reload.go b/plugin/reload/reload.go
index e469b452..35b41997 100644
--- a/plugin/reload/reload.go
+++ b/plugin/reload/reload.go
@@ -1,20 +1,21 @@
+// Package reload periodically checks if the Corefile has changed, and sends the reload signal
+// signal (SIGUSR1) to itself to reload.
 package reload
 
 import (
 	"crypto/md5"
 	"log"
+	"os"
+	"syscall"
 	"time"
 
 	"github.com/mholt/caddy"
 )
 
-// reload periodically checks if the Corefile has changed, and reloads if so
-
 type reload struct {
 	instance *caddy.Instance
 	interval time.Duration
 	sum      [md5.Size]byte
-	stopped  bool
 	quit     chan bool
 }
 
@@ -23,13 +24,6 @@ func hook(event caddy.EventName, info interface{}) error {
 		return nil
 	}
 
-	// if reload is removed from the Corefile, then the hook
-	// is still registered but setup is never called again
-	// so we need a flag to tell us not to reload
-	if r.stopped {
-		return nil
-	}
-
 	// this should be an instance. ok to panic if not
 	r.instance = info.(*caddy.Instance)
 	r.sum = md5.Sum(r.instance.Caddyfile().Body())
@@ -46,13 +40,12 @@ func hook(event caddy.EventName, info interface{}) error {
 				}
 				s := md5.Sum(corefile.Body())
 				if s != r.sum {
-					_, err := r.instance.Restart(corefile)
+					pid := os.Getpid()
+					err := syscall.Kill(pid, syscall.SIGUSR1)
 					if err != nil {
-						log.Printf("[ERROR] Corefile changed but reload failed: %s\n", err)
+						log.Printf("[ERROR] Corefile changed but failed to sent SIGUSR1 to %d: %s\n", pid, err)
 						continue
 					}
-					// we are done, this hook gets called again with new instance
-					r.stopped = true
 					return
 				}
 			case <-r.quit:
~~~

**Comment by fturib** [comment]
I think that the issue is not really on the Restart itself.
So you way to do the restart should work.


Issue is more that in the loop to loadCqaddyFile you depend of the value "r.instance" that is a global var.

When you restart, at setup time RELOAD will re-initialized this r variable with a null instance.
if during the same setup, a directive that is processed AFTER reload is failing .. then r will stay with a null instance, although the server is not restarted (beacause of failing setup.)

on next loop, you will use r.instance to load your corefile . .. but r.instance is NIL.


from there my wondering about global vars : some can be recomputed during the setup process, but that new setup may not be globally successful and the service will continue (without restarting), but with some global values that may not correspond to the right corefile.

**Comment by fturib** [comment]
I can reproduce the case using this corefile:

```
.:5354 {
     errors
     health :8091
     proxy . 2.2.2.2:53 {
        fail_timeout 9s
        max_fails 1
     }
     log
     reload 2s
}
```

run coredns -> all is ok.

```
API server listening at: 127.0.0.1:51513
.:5354
2018/01/28 14:02:00 [INFO] CoreDNS-1.0.5
2018/01/28 14:02:00 [INFO] darwin/amd64, go1.9.2, 
CoreDNS-1.0.5
darwin/amd64, go1.9.2, 
```

then modify the corefile, changing "max_fails 1" to "max_fails 1s" and save the corefile

```
2018/01/28 14:02:05 [INFO] Reloading
2018/01/28 14:02:05 [ERROR] Corefile changed but reload failed: plugin/proxy: strconv.Atoi: parsing "1s": invalid syntax
```

then try to change back to : "max_fails 1"
=> Nothing happen, the loop of the hook is lost ...

running directly on mac (I mean without the debugger), I get:
(in fact the r.instance is nil)

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1eb47c2]

goroutine 76 [running]:
github.com/mholt/caddy.(*Instance).Caddyfile(...)
        /Users/ftur/go/src/github.com/mholt/caddy/caddy.go:448
github.com/coredns/coredns/plugin/reload.hook.func1()
        /Users/ftur/go/src/github.com/coredns/coredns/plugin/reload/reload.go:43 +0x132
created by github.com/coredns/coredns/plugin/reload.hook
        /Users/ftur/go/src/github.com/coredns/coredns/plugin/reload/reload.go:37 +0x12b
ftur-MBP13:coredns ftur$ 
```

**Comment by fturib** [comment]
In fact, this plugin RELOAD is VERY useful, especially if you try to tune your corefile, changing parameters to observe the behavior ...
That is how I found the issue.

**Comment by fturib** [comment]
NOTE: I may have misplaced the 'r.stopped = true'. - need to rethink of it.

@miekg : if we use the SIGUSR1 .. will it work on Windows ? - I am thinking 'no' because SIGUSR1 is posix signal.

**Comment by miekg** [comment]
Graceful reload doesn't work on windows. But good point. Sending signals
opens that can of worms.

File: plugin/reload/setup.go [hunk 1]
```diff
@@ -32,19 +37,25 @@ func setup(c *caddy.Controller) error {
 	if len(args) > 0 {
 		d, err := time.ParseDuration(args[0])
 		if err != nil {
-			return err
+			return plugin.Error("reload", err)
 		}
 		i = d
 	}
+	if i < minInterval {
+		return plugin.Error("reload", errors.New("interval value must be greater or equal to 2 sec"))
```
Comment (johnbelamaric) [inline plugin/reload/setup.go]: fmt.Errorf("interval values must be greater or equal to %v", minInterval)

File: plugin/reload/setup.go [hunk 2]
```diff
@@ -32,19 +37,25 @@ func setup(c *caddy.Controller) error {
...
 	if len(args) > 1 {
 		d, err := time.ParseDuration(args[1])
 		if err != nil {
-			return err
+			return plugin.Error("reload", err)
 		}
 		j = d
 	}
+	if j < minJitter {
+		return plugin.Error("reload", errors.New("jitter value must be greater or equal to 1 sec"))
```
Comment (johnbelamaric) [inline plugin/reload/setup.go]: ditto

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Ok, this makes sense. Minor nit.

**Comment by johnbelamaric** [comment]
By the way, since this has some odd edge conditions, you should build integration tests for it. You can watch the log to see if the right reloads happen...

**Comment by miekg** [comment]
can we also unit test this?

**Comment by johnbelamaric** [comment]
I think it's a bit tricky to unit test because it involves server restarts, etc. But maybe.

By the way, the issue with package-level vars in other plugins that this issue raises applies to signal-based restarts too. There is no real difference. We probably want to review all of those, as I am not sure they were all written with the idea of `setup` being called more than once.

**Review by johnbelamaric — APPROVED** [review]

**Comment by johnbelamaric** [comment]
Additional integration tests coming in another PR.
